Feature/ct 426 scratch

bneijt opened the pull request:
Scratch PR to point share some code information and point to interesting code lines to create an initial draft for https://github.com/dbt-labs/dbt-core/issues/4957

McKnight-42 committed:
include directory README (#4685)

* start of a README for the include directory

* minor updates

* minor updates after comments from gerda and emily

* trailing space issue?

* black formatting

* minor word change

* typo update

* minor fixes and changelog creation

* remove changelog

agoblet committed:
add DO_NOT_TRACK environment variable support (#5000)

McKnight-42 committed:
init push up of converted unique_key tests (#4958)

* init push up of converted unique_key tests

* testing cause of failure

* adding changelog entry

* moving non basic test up one directory to be more broadly part of adapter zone

* minor changes to the bad_unique_key tests

* removed unused fixture

* moving tests to base class and inheriting in a simple class

* taking in chenyu's changes to fixtures

* remove older test_unique_key tests

* removed commented out code

* uncommenting seed_count

* v2 based on feedback for base version of testing, plus small removal of leftover breakpoint

* create incremental test directory in adapter zone

* commenting out TableComparision and trying to implement check_relations_equal instead

* remove unused commented out code

* changing cast for date to fix test to work on bigquery

gshank committed:
Remove TableComparison and convert existing calls to use dbt.tests.util (#4986)

iknox-fa committed:
updated event message (#5011)

ChenyuLInx committed:
adapter_methods (#4939)

* adapter_methods

* fix fixture scope

* update table compare method

* remove unneeded part

* update test name and comment

ChenyuLInx committed:
make graph_selection tests just checking selection (#5012)

* make graph_selection tests just checking selection

* use util function

iknox-fa committed:
added git-blame-ignore-revs file (#5019)

iknox-fa committed:
Decouple project creation logic from tasks CT-299 (#4981)

iknox-fa committed:
updated docker readme CT-452 (#5018)

dependabot[bot] committed:
Bump black from 22.1.0 to 22.3.0 (#4972)

* Bump black from 22.1.0 to 22.3.0

gshank committed:
Fix hard_delete_snapshot test to do the right thing. (#5020)

stu-k committed:
Convert --version tests to pytest (#5026)

Convert --version tests to pytest

emmyoop committed:
Update index file for docs generation (#4995)

* Update index file for docs generation

* add changelog entries

poloaraujo committed:
Added no-print flag (#4854)

* Added no-print flag

* Updated changelog

* Updated changelog

* Removed changes from CHANGELOG.md

* Updated CHANGELOG.MD with changie

* Update .changes/unreleased/Features-20220408-114118.yaml

Co-authored-by: Emily Rockman <ebuschang@gmail.com>

Co-authored-by: Emily Rockman <ebuschang@gmail.com>

emmyoop committed:
cache after retrying instead of while retrying (#5028)

sungchun12 committed:
Smart Source Freshness Runs (#4256)

* first draft

* working selector code

* remove debug print logs

* copy test template

* add todo

* smarter depends on graph searching notes

* add excluded source children nodes

* remove prints and clean up logger

* opinionated fresh node selection

* better if handling

* include logs with meaningul info

* add concurrent selectors note

* cleaner logging

* Revert "Merge branch 'main' of https://github.com/dbt-labs/dbt into feature/smart-source-freshness-runs"

This reverts commit 7fee4d44bf0c5da9ccfaf01634b823f8786bf2bf, reversing
changes made to 17c47ff42d79f0729e6fdfe7259e33eb65cb127a.

* tidy up logs

* remove comments

* handle criterion that does not match nodes

* use a blank set instead

* Revert "Revert "Merge branch 'main' of https://github.com/dbt-labs/dbt into feature/smart-source-freshness-runs""

This reverts commit 71125167a1d00979c3e333530b5e5d196304468c.

* make compatible with rc and new logger

* new log format

* new selector flag name

* clarify that status needs to be correct

* compare current and previous state

* correct import

* add current state

* remove print

* add todo

* fix error conditions

* clearer refresh language

* don't run wasteful logs

* remove for now

* cleaner syntax

* turn generator into set

* remove print

* add fresh selector

* data bookmarks matter only

* remove exclusion logic for status

* keep it DRY

* remove unused import

* dynamic project root

* dynamic cwd

* add TODO

* simple profiles_dir import

* add default target path

* headless path utils

* draft work

* add previous sources artifact read

* make PreviousState aware of t-2 sources

* make SourceFreshSelectorMethod aware of t-2 sources

* add archive_path() for t-2 sources to freshness.py

* clean up merged branches

* add to changelog

* rename file

* remove files

* remove archive path logic

* add in currentstate and previousstate defaults

* working version of source fresher

* syntax source_fresher: works

* fix quoting

* working version of target_path default

* None default to sources_current

* updated source selection semantics

* remove todo

* move to test_sources folder

* copy over baseline source freshness tests

* clean up

* remove test file

* update state with version checks

* fix flake tests

* add changelog

* fix name

* add base test template

* delegate tests

* add basic test to ensure nothing runs

* add another basic test

* fix test with copy state

* run error test

* run warn test

* run pass test

* error handling for runtime error in source freshness

* error handling for runtime error in source freshness

* add back fresher selector condition

* top level selector condition

* add runtime error test

* testing source fresher test selection methods

* fix formatting issues

* fix broken tests

* remove old comments

* fix regressions in other tests

* add Anais test cases

* result selector test case

Co-authored-by: Matt Winkler <matt.winkler@fishtownanalytics.com>

leahwicz committed:
Updating tests and doc to support Python 3.10 (#5025)

* Updating tests and doc to support Python 3.10

* Single quotes needed for python version matrix

* Adding changelog

bneijt committed:
Add config and scratch run method

bneijt committed:
Lock dependencies using poetry

bneijt commented on core/dbt/contracts/project.py:
```diff
@@ -263,7 +263,7 @@ class ProfileConfig(HyphenatedDbtClassMixin, Replaceable):
     # TODO: make this a dynamic union of some kind?
     credentials: Optional[Dict[str, Any]]
-
+    manage_schema: Optional[bool]
```
This merges the manage_schema configuration with the profile, but I think it might be nicer to move this configuration to a top-level `manage_targets` configuration (list of targets to enable schema management for).
The difference is having
```
config:
  manage_targets: [test, prod]
```
vs
```
snowflake:
  outputs:
    dev:
      manage_schema: true
```

Not sure about this yet, I also don't know if target names are unique per driver or per profile config file...

agoblet replied:
If I understand this correct, your code change is done in project, allowing the user to specify this in `dbt_project.yml` eventually, while your comment talks about the profile. 
- Do I understand correctly?
- I think doing this on a project level makes it no longer possible to have different configs for dev and pro, for example.

agoblet replied:
The second option you mention in your comment is clearer to me. When looking at dev, you know its config right away, without moving to another place in the yaml to check the manage_targets variable.

What pros do you see with your first option?

bneijt commented on core/dbt/task/run.py:
```diff
@@ -483,3 +484,30 @@ def get_runner_type(self, _):
+        # Find a list of relations in a given schema
+        # List everything in the schema
+        # Drop the difference
+        print("TYPE IS", type(adapter))
```
I really need a nice working debug workflow or isolated test to get this working properly :D

emmyoop committed:
Add enabled as a source config (#5008)

* initial pass at source config test w/o overrides

* Update tests/functional/sources/test_source_configs.py

Co-authored-by: Jeremy Cohen <jeremy@dbtlabs.com>

* Update tests/functional/sources/test_source_configs.py

Co-authored-by: Jeremy Cohen <jeremy@dbtlabs.com>

* tweaks from feedback

* clean up some test logic - add override tests

* add new fields to source config class

* fix odd formatting

* got a test working

* removed unused tests

* removed extra fields from SourceConfig class

* fixed next failing unit test

* adding back missing import

* first pass at adding source table configs

* updated remaining tests to pass

* remove source override tests

* add comment for config merging

* changelog

* remove old comments

* hacky fix for permission test

* remove unhelpful test

* adding back test file that was accidentally deleted

Co-authored-by: Jeremy Cohen <jeremy@dbtlabs.com>
Co-authored-by: Nathaniel May <nathaniel.may@fishtownanalytics.com>
Co-authored-by: Chenyu Li <chenyu.li@dbtlabs.com>

emmyoop committed:
add more complete logic around changelog contributors section (#5037)

* add more complete logic around changelog contributors section

* add instructions for future core team members

* Update .changie.yaml

jtcohen6 committed:
Add experimental `cache_selected_only` config (#5036)

* cache schema for selected models

* Create Features-20220316-003847.yaml

* rename flag, update postgres adapter

rename flag to cache_selected_only, update postgres adapter: function _relations_cache_for_schemas

* Update Features-20220316-003847.yaml

* added test for cache_selected_only flag

* formatted as per pre-commit

* Add breaking change note for adapter plugin maintainers

* Fix whitespace

* Add a test

Co-authored-by: karunpoudel-chr <poudel.karun@gmail.com>
Co-authored-by: karunpoudel-chr <62040859+karunpoudel@users.noreply.github.com>

b-per committed:
Add `selected_resources` to the Jinja context (#5001)

* Add selected_resources in the Jinja context

* Add tests for the Jinja variable selected_resources

* Add Changie entry for the addition of selected_resources

* Move variable to the ProviderContext

* Move selected_resources from ModelContext to ProviderContext

* Update unit tests for context to cater for the new selected_resources variable

* Move tests to a Class where tests are run after a dbt build

stu-k committed:
Update version output and logic (#5029)

Update version output and logic

Commit:
Perf regression testing - overhaul of readme and runner (#4602)

leahwicz committed:
Bumping manifest schema to v5 (#5032)

* Bumping manifest schema to v5

* Adding changelog

github-actions[bot] committed:
Bumping version to 1.2.0a1 (#5045)

* Bumping version to 1.2.0a1

* Fixing spacing issue

Co-authored-by: Github Build Bot <buildbot@fishtownanalytics.com>
Co-authored-by: leahwicz <60146280+leahwicz@users.noreply.github.com>

adamantike committed:
Flexibilize MarkupSafe pinned version (#5039)

* Flexibilize MarkupSafe pinned version

The current `MarkupSafe` pinned version has been added in #4746 as a
temporary fix for #4745.

However, the current restrictive approach isn't compatible with other
libraries that could require an even older version of `MarkupSafe`, like
Airflow `2.2.2` [0], which requires `markupsafe>=1.1.1, <2.0`.

To avoid that issue, we can allow a greater range of supported
`MarkupSafe` versions. Considering the direct dependency `dbt-core` has
is `Jinja2==2.11.3`, we can use its pinning as the lower bound, which is
`MarkupSafe>=0.23` [1].

This fix should be also backported this to `1.0.latest` for inclusion in
the next v1.0 patch.

[0] https://github.com/adamantike/airflow/blob/2.2.2/setup.cfg#L125
[1] https://github.com/pallets/jinja/blob/2.11.3/setup.py#L53

amy-byrum committed:
Add updated dbt diagram for readme (#5055)

joellabes committed:
Clarify steps to reopen a stale issue (#4802)

I haven't added a message for stale PRs because they're likely to only impact the opening user (who I assume can reopen their own PR) and they're less of a problem. Happy to add that in as well, as well as to take feedback on the specific phrasing here.

emmyoop committed:
move deprecation check outside package caching (#5069)

* move deprecation check outside package caching

* add changelog

VersusFacit committed:
Ct 488/migrate simple seed (#5060)

* (finally) idiomatically rewrite a class of tests into the new framework.

* Get simple seed mostly working with design tweaks needed.

* Revamp tests to use more of the framework. Fix TODOs

* Complete migration of 005 and remove old files.

* Fix BOM test for Windows and add changelog entry

* Finalize tests in the adapter zone per conversation with Chenyu.

Co-authored-by: Mila Page <versusfacit@users.noreply.github.com>

Commit:
add managed schema configuraiton to project spec

Commit:
add base classes to SchemaManagementConfiguration

Commit:
change order of class definitions in projectt

Commit:
change schema management action to literal type

bneijt committed:
Broken config impl

Commit:
change schema management action to str type

bneijt committed:
code running

bneijt committed:
Always fail

bneijt committed:
Drop relations logic

bneijt committed:
Add config

bneijt committed:
Add management logic

bneijt committed:
Add pre-commit

bneijt committed:
Poetry update

bneijt committed:
Black formatting and flake results, fix logic

leahwicz committed:
Updating backport action to latest (#5082)

* Updating backport action to latest

* Updating to PR trigger with permissions instead

This is a better model for closing down all permissions and just granting what we actually want

* Updating IF when merged and backport label exists

* Changing to only trigger on label being added

emmyoop committed:
CT-476 convert deprecation tests (#5034)

* first pass at 012 test conversion

* convert 012_deprecation tests

* add logic around dropping schema

* swap exception

* added clarifying comment

emmyoop committed:
cleanup changelog files on main after cutting release branch (#5112)

* cleanup cahngelog files on main after release branch was cut

* last file that is backported

gshank committed:
Convert docs_generate_tests to new framework (#5058)

emmyoop committed:
add new GHA for dependabot PRs (#5065)

* add new GHA for dependabot PRs

* Add automated changelog entry

* code cleanup

* remove changelog file

* permissions tweak

* Add automated changelog yaml from template

* update commit author

* Add automated changelog yaml from template

* fix formatting, remove changelog

* revert to separate files and comment out changelog check temporarily

* Add automated changelog yaml from template

* add back changelog check, update how commit works

* remove file

* Add automated changelog yaml from template

* WIP update to use PAT

* update PAT name

* remove file

* Add automated changelog yaml from template

* format file with quotes

* delete file

* Add automated changelog yaml from template

* remove extra line

* remove file

* Add automated changelog yaml from template

* Delete Dependencies-20220418-194629.yaml

Co-authored-by: Github Build Bot <buildbot@fishtownanalytics.com>

dbeatty10 committed:
Restore ability to utilize `updated_at` for check_cols snapshots (#5077)

* Restore ability to configure and utilize `updated_at` for snapshots using the check_cols strategy

* Changelog entry

* Optional comparison of column names starting with `dbt_`

* Functional test for check cols snapshots using `updated_at`

* Comments to explain the test implementation

jtcohen6 committed:
GH hygiene: contributing guide, templates, stalebot (#4967)

* Update contributing guide

* Update issue + PR templates

* Stalebot for all issues, no exceptions

* Update links

* Missed one

* PR feedback

emmyoop committed:
fix retry logic failures (#5137)

* fix retry logic failures

* changelog

* add tests to make sure data is getting where it needs to

* rename file

* remove duplicate file

leahwicz committed:
Updating backport Action permissions (#5121)

jtcohen6 committed:
Even more scrubbing (#5152)

* Even more scrubbing

* Changelog entry

* Even more

* remove reduendent scrub

* remove reduendent scrub

* fix encoding issue

* keep scrubbed log in args

Co-authored-by: Chenyu Li <chenyu.li@dbtlabs.com>

gshank committed:
Use yaml renderer (with target context) for rendering selectors (#5136)

* Use yaml renderer (with target context) for rendering selectors

* Changie

* Convert cli_vars tests

* Add test for var in profiles

* Add test for cli vars in packages

* Add test for vars in selectors

emmyoop committed:
add triage label workflow (#5164)

emmyoop committed:
Tweak `triage-label.yml` to trigger off issue labels instead of PR labels (#5168)

* fix label check

* fix label filter to go against issues not PRs

VersusFacit committed:
Ct 488/migrate invalid model tests (#5143)

* First test completed.
* Convert and update more test cases.
* Complete test migration and remove old files.

Co-authored-by: Mila Page <versusfacit@users.noreply.github.com>

bd3dowling committed:
Add itertools to modules (#5140)

* GH hygiene: contributing guide, templates, stalebot (#4967)

* Update contributing guide

* Update issue + PR templates

* Stalebot for all issues, no exceptions

* Update links

* Missed one

* PR feedback

* Update CHANGELOG

Co-authored-by: Jeremy Cohen <jeremy@dbtlabs.com>

VersusFacit committed:
migrate 009_data_tests to new test framework (#5139)

* Fold so-called 'data' test into new framework with new vocabulary to match.
* Add missing files including changelog.
* Remove unneeded Changelog per team policy on test conversions.
* Refactor test code to better use our pytest framework. Strengthen assertions.

Co-authored-by: Mila Page <versusfacit@users.noreply.github.com>

ChenyuLInx committed:
remove extra class and add connection test (#5163)

* remove extra class and add connection test

* add project artifact to avoid breaking other tests

* add comment

danieldiamond committed:
Add selector method capabilities to selectors (#4827)

GtheSheep committed:
Fix adding new cols to check_cols in snapshots (#4893)

gshank committed:
Ct-65 metrics names with spaces (#5173)

* Convert existing metrics test

* add non-failing test for names with spaces

* Raise ParsingException if metrics name contains spaces

* Remove old metrics tests

gshank committed:
When parsing 'all_sources' should be a list of unique dirs (#5176)

* When parsing 'all_sources' should be a list of unique dirs

* Changie

* Fix some unit tests of all_source_paths

* Convert 039_config_tests

* Remove old 039_config_tests

* Add test for duplicate directories in 'all_source_files'

jtcohen6 committed:
Update README for dbt-tests-adapter (#5182)

* Update README for dbt-tests-adapter

* Add logo

jeremyyeo committed:
Fix: add warning on duplicated yaml keys (#5146)

* add warning on duplicated yaml keys

* update structure and tests

* fix old test schema file

* add changelog

stu-k committed:
Convert list tests to pytest (#5178)

dependabot[bot] committed:
Bump ubuntu from 20.04 to 22.04 (#5141)

* Bump ubuntu from 20.04 to 22.04

emmyoop committed:
remove reference to unused `ok to test` label (#5149)

leahwicz committed:
Adding Skip Changelog label to Version Bump action (#5203)

emmyoop committed:
update label check (#5194)

gshank committed:
Convert 013_context_vars_tests to context_methods. Move existing cli_vars into context_methods (#5199)

adamantike committed:
seed: Add new macro get_csv_sql (#5207)

new macro `get_csv_sql`

alexrosenfeld10 committed:
Allow the target name to be set in profile_template.yml (#5184)

* Allow the target name to be set in profile_template.yml

McKnight-42 committed:
propseal for modification to drop_test_schema (#5198)

* propseal for modification to drop_test_schema

* changelog

* remove hard coded run_dbt version and put back previous version of drop_test_schema, add commit to drop_schema

NicolasPA committed:
Improve tracking error logging message (#5211)

gshank committed:
Convert materialization tests to new test framework (#5212)

iknox-fa committed:
Update mypy to latest and turn on everywhere (#5171)

iknox-fa committed:
Revert inaccurate mypy change (#5216)

leahwicz committed:
Bumping hologram version (#5218)

* Bumping hologram version

* Add automated changelog yaml from template

* Updating issue

* Loosen requirement range

Co-authored-by: Jeremy Cohen <jeremy@dbtlabs.com>

Co-authored-by: Github Build Bot <buildbot@fishtownanalytics.com>
Co-authored-by: Jeremy Cohen <jeremy@dbtlabs.com>

jtcohen6 committed:
Fix: column comparison logic for check-strategy snapshots (#5223)

* Add test case

* Update comparison in snapshot_check_all_get_existing_columns

* Add changelog entry

jtcohen6 committed:
Fix: use adapter.get_columns_in_relation (#5232)

jeremyyeo committed:
Feature: Add set/zip function to contexts (#5107)

* add set function to contexts

* add zip function to contexts

* add changelog

* add try_ equivalents

* remove defaults

* add tests

* update tests

tomasfarias committed:
fix: Avoid access to profile when calling str(UnsetProfileConfig) (#5209)

* fix: Avoid access to profile when calling str(UnsetProfileConfig)

dbt.config.UnsetProfileConfig inherits __str__ from
dbt.config.Project. Moreover, UnsetProfileConfig also raises an
exception when attempting to access unset profile attributes. As
Project.__str__ ultimately calls to_project_config and accesses said
profile attributes, we override to_project_config in
UnsetProfileConfig to avoid accessing the attributes that raise an
exception.

This allows calling str(UnsetProfileConfig) and
repr(UnsetProfileConfig).

Basic unit testing is also included in commit.

* fix: Skip repr for profile fields in UnsetProfileConfig

* chore(changie): Add changie file

Commit:
Add support for File Selectors and add file selectors to the default method selector list (#5241)

* Add a new selector method for files and add it to the default method selection criteria if the given selector has a . in it but no path separators

* Add a file: selector method to the default selector methods because it will make Pedram happy

* changie stuff

stu-k committed:
Fix macro modified from previous state with pkg (#5224)

* Fix macro modified from previous state with pkg

When iterating through nodes to check if any of its macro dependencies
have been modified, the state selector will first check all upstream
macro dependencies before returning a judgement.

jtcohen6 committed:
Add dbt Core roadmap as of May 2022 (#5246)

gshank committed:
Tweak test to avoid set ordering problem (#5272)

leahwicz committed:
Creating ADR for versioning and branching strategy (#4998)

* Creating ADR for versioning and branching strategy

* Fixing image link

* Grammar clean-up

Co-authored-by: Stu Kilgore <stuart.kilgore@gmail.com>

* Grammar clean-up

Co-authored-by: Stu Kilgore <stuart.kilgore@gmail.com>

* Update docs/arch/adr-003-versioning-branching-strategy.md

Co-authored-by: Stu Kilgore <stuart.kilgore@gmail.com>

* Update docs/arch/adr-003-versioning-branching-strategy.md

Co-authored-by: Stu Kilgore <stuart.kilgore@gmail.com>

* Update docs/arch/adr-003-versioning-branching-strategy.md

Co-authored-by: Stu Kilgore <stuart.kilgore@gmail.com>

* Update docs/arch/adr-003-versioning-branching-strategy.md

Co-authored-by: Stu Kilgore <stuart.kilgore@gmail.com>

* Update docs/arch/adr-003-versioning-branching-strategy.md

Co-authored-by: Stu Kilgore <stuart.kilgore@gmail.com>

* Update docs/arch/adr-003-versioning-branching-strategy.md

Co-authored-by: Stu Kilgore <stuart.kilgore@gmail.com>

* Update docs/arch/adr-003-versioning-branching-strategy.md

Co-authored-by: Stu Kilgore <stuart.kilgore@gmail.com>

* Update docs/arch/adr-003-versioning-branching-strategy.md

Co-authored-by: Stu Kilgore <stuart.kilgore@gmail.com>

* Updating Outside Scope section

* Changing from using type to stage

* Adding section on getting changes into certain releases

* Changed stages to phases

* Some wording updates

* New section for branching pros and cons

* Clarifying version bump statement

* A few minor comment fix ups

* Adding requirement to define released

* Updating to completed!

Co-authored-by: Stu Kilgore <stuart.kilgore@gmail.com>

epapineau committed:
Truncate relation names when appending a suffix (#4921)

* Truncate relation names when appending a suffix that will result in len > 63 characters using make_temp_relation and make_backup_relation macros

* Remove timestamp from suffix appended to backup relation

* Add changelog entry

* Implememt make_relation_with_suffix macro

* Add make_intermediate_relation macro that controls _tmp relation creation in table and view materializations to delienate from database- and schema-less behavior of relation returned from make_temp_relation

* Create backup_relation at top of materialization to use for identifier

* cleanup

* Add dstring arg to make_relation_with_suffix macro

* Only reference dstring in conditional of make_relation_with_suffix macro

* Create both a temp and intermediate relation, update preexisting_temp_relation to preexisting_intermediate_relation

* Migrate test updates to new test location

* Remove restored tmp.csv

* Revert "Remove restored tmp.csv"

This reverts commit 900c9dbcad9a1e6a5a6737c84004504bfdd9926f.

* Actually remove restored tmp.csv

gshank committed:
CT-159 Remove docs file from manifest when deleting doc node (#5270)

* Convert partial parsing docs tests

* Failing test

* Remove doc file from manifest when doc node is removed

* Changie

Commit:
Switch the Makefile to use the default installed Python version instead of py38 for local testing (#5269)

* Setting up an env var to use to override the tox python variable used for local dev

* Switch over to py-based tox envs instead of the py38 ones to be friendly to dbt-core devs who don't work at dbt Labs

* changie

gshank committed:
Ct-581 grants as configs (#5230)

* Handle 'grants' in NodeConfig, with correct merge behavior

* Fix a bunch of tests

* Add changie

* Actually add the test

* Change to default replace of grants with '+' extending them

* Additional tests, fix config_call_dict handling

* Tweak _add_config_call to remove unnecessary isinstance checks

bneijt committed:
Start tests folder

bneijt committed:
Add update_models utility method on project info

bneijt committed:
Move configuration to credentials class

Commit:
add framework integration test

Commit:
fix profile config

bneijt committed:
Move schemas to snapshot_paths

Commit:
fix managed schemas project

bneijt committed:
Fix schema validation error

bneijt committed:
Fix manage_schemas parsing from profile

bneijt committed:
Happy flow

Commit:
add noop and warn tests

Commit:
improve tests
